File: game.java
package com.cardgame.model;

import java.util.List;

public class CardGame {
    public enum Suit {
        CLUB, DIAMOND, SPADE, HEART
    }

    public enum Rank {
        TWO, THREE, FOUR, FIVE, SIX, SEVEN, EIGHT, NINE, TEN,
        JACK, QUEEN, KING, ACE
    }

    public record Card(Suit suit, Rank rank) {}

    public record Player(String name, List<Card> hand) {}

    public record Game(List<Card> deck, List<Player> players) {}

    @FunctionalInterface
    public interface Deal {
        DealResult apply(List<Card> deck);
    }

    public record DealResult(List<Card> remainingDeck, Card dealtCard) {}

    @FunctionalInterface
    public interface PickupCard {
        List<Card> apply(List<Card> hand, Card card);
    }
}

File: game.zig
pub const Suit = enum {
    club,
    diamond,
    spade,
    heart,
};

pub const Rank = enum {
    two,
    three,
    four,
    five,
    six,
    seven,
    eight,
    nine,
    ten,
    jack,
    queen,
    king,
    ace,
};

pub const Card = struct {
    suit: Suit,
    rank: Rank,
};

pub const Hand = []const Card;
pub const Deck = []const Card;

pub const Player = struct {
    name: []const u8,
    hand: Hand,
};

pub const Game = struct {
    deck: Deck,
    players: []const Player,
};

pub const DealResult = struct {
    deck: Deck,
    card: Card,
};

pub const Deal = *const fn (Deck) DealResult;
pub const PickupCard = *const fn (Hand, Card) Hand;
